fv3config/config/diag_table.py

```diff
@@ -0,0 +1,212 @@
+import logging
+from typing import Sequence, Optional, Union
+import dataclasses
+import datetime
+import re
+
+import dacite
+
+from .._exceptions import ConfigError
+
+logger = logging.getLogger("fv3config")
+
+
+@dataclasses.dataclass
+class DiagTableField:
+    """Object representing a field of a diagnostics file."""
+
+    module_name: str
+    field_name: str
+    output_name: str
+    reduction_method: Optional[str] = "none"
+    time_sampling: Optional[str] = "all"
+    regional_section: Optional[str] = "none"
+    packing: Optional[int] = 2
+
+
+@dataclasses.dataclass
+class DiagTableFile:
+    """Object representing a diagnostics file."""
+
+    name: str
+    frequency: int
+    frequency_units: str
+    fields: Sequence[DiagTableField]
+    file_format: Optional[int] = 1
+    time_axis_units: Optional[str] = "hours"
+    time_axis_name: Optional[str] = "time"
+
+
+class DiagTable:
+    def __init__(
+        self, name: str, base_time: datetime.datetime, files: Sequence[DiagTableFile],
+    ):
+        """Representation of diag_table, which controls Fortran diagnostics manager.
+
+        Note:
+            This implementation is based on the diag_table specification described in
+      https://data1.gfdl.noaa.gov/summer-school/Lectures/July16/03_Seth1_DiagManager.pdf
+            The MOM6 documentation has a useful description as well:
+            https://mom6.readthedocs.io/en/latest/api/generated/pages/Diagnostics.html.
+
+        Args:
+            name: label used as attribute in output diagnostic files. Cannot contain
+                spaces.
+            base_time: time to be used as reference for time coordinate units.
+            files: sequence of DiagTableFile's defining the diagnostics to be output.
+        """
+        if " " in name:
+            raise ConfigError(f"Name for diag_table cannot have spaces. Got '{name}'.")
+        self.name = name
+        self.base_time = base_time
+        self.files = files
+
+    def __repr__(self):
+        """Representation of diag_table expected by the Fortran model."""
+        lines = []
+        lines.append(self.name)
+        lines.append(self._time_to_str(self.base_time))
+        lines.append("")
+
+        for file_ in self.files:
+            lines.append(self._file_repr(file_))
+
+        lines.append("")
+        for file_ in self.files:
+            for field in file_.fields:
+                lines.append(self._field_repr(field, file_.name))
+            lines.append("")
+
+        return "\n".join(lines)
+
+    def asdict(self):
```

Comment:
I know this is mirroring dataclasses, but since we also have `from_dict` with pothole case can this use pothole case for consistency with itself/the rest of our user code?

Reply:
I would prefer to be consistent with the standard library here, although I agree it's a bit unsatisfying.